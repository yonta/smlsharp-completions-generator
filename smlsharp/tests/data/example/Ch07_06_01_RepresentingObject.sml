fn self => fn x => self := (!self # {X = x});

val pointClass =
{
  getX = fn self => #X (!self),
  setX = fn self => fn x => self := (!self # {X = x}),
  getY = fn self => #Y (!self),
  setY = fn self => fn x => self := (!self # {Y = x}),
  getColor = fn self => #Color (!self),
  setColor = fn self => fn x => self := (!self # {Color = x})
};

local
  val state = ref { X = 0.0, Y = 0.0 }
in 
  val myPoint = fn method => method pointClass state
end;

local
  val state = ref { X = 0.0, Y = 0.0, Color = "Red" }
in 
  val myColorPoint = fn method => method pointClass state
end;

myPoint # setX 1.0;
myPoint # getX;
myColorPoint # getX;
myColorPoint # getColor;
myPoint # getColor;
