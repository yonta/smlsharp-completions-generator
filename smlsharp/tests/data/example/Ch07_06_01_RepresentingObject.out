val it = fn : ['a#{X: 'b}, 'b. 'a ref -> 'b -> unit]

val pointClass =
  {getColor = fn, getX = fn, getY = fn, setColor = fn, setX = fn, setY = fn}
  : {getColor: ['a#{Color: 'b}, 'b. 'a ref -> 'b],
     getX: ['a#{X: 'b}, 'b. 'a ref -> 'b],
     getY: ['a#{Y: 'b}, 'b. 'a ref -> 'b],
     setColor: ['a#{Color: 'b}, 'b. 'a ref -> 'b -> unit],
     setX: ['a#{X: 'b}, 'b. 'a ref -> 'b -> unit],
     setY: ['a#{Y: 'b}, 'b. 'a ref -> 'b -> unit]}

val myPoint =
  fn
  : ['a#{Color: 'b},
     'b,
     'c#{X: 'd},
     'd,
     'e#{Y: 'f},
     'f,
     'g#{Color: 'h},
     'h,
     'i#{X: 'j},
     'j,
     'k#{Y: 'l},
     'l,
     'm.
     ({getColor: 'a ref -> 'b,
       getX: 'c ref -> 'd,
       getY: 'e ref -> 'f,
       setColor: 'g ref -> 'h -> unit,
       setX: 'i ref -> 'j -> unit,
       setY: 'k ref -> 'l -> unit}
      -> {X: real, Y: real} ref -> 'm)
     -> 'm]

val myColorPoint =
  fn
  : ['a#{Color: 'b},
     'b,
     'c#{X: 'd},
     'd,
     'e#{Y: 'f},
     'f,
     'g#{Color: 'h},
     'h,
     'i#{X: 'j},
     'j,
     'k#{Y: 'l},
     'l,
     'm.
     ({getColor: 'a ref -> 'b,
       getX: 'c ref -> 'd,
       getY: 'e ref -> 'f,
       setColor: 'g ref -> 'h -> unit,
       setX: 'i ref -> 'j -> unit,
       setY: 'k ref -> 'l -> unit}
      -> {Color: string, X: real, Y: real} ref -> 'm)
     -> 'm]

val it = () : unit

val it = 1.0 : real

val it = 0.0 : real

val it = "Red" : string
